### PR TITLE
[low-code CDK] Rsumable full refresh support for low-code streams

### PR DESCRIPTION
**Pull request by brianjlai**
Closes https://github.com/airbytehq/airbyte-internal-issues/issues/7000 (nice issue id)

## What

https://www.loom.com/share/97b0ee3b050448aba2cbfdd5b7d91e3d

This PR makes all low-code streams support resumable full refresh if they implement a paginator and are not already implemented as incremental syncs. 

This approach should work with substreams since the new cursor adheres to the same interface as DatetimeBasedCursor which can be used by a per-partition cursor, but I've explicitly gated it off to reduce the scope.

## How

At a high level the changes are:

- A new checkpoint reader that uses a stream's underlying Cursor implementation to manage how state is persisted and accessed
- A new `ResumableFullRefreshCursor` which adheres to the Cursor interface
- Updating all pagination strategies to support optionally reseting based on an incoming page/offset/cursor
- Updating the simple retriever to allow for reading single pages and the flow for restarting a sync from an incoming page state
- Updating the model to component transformer to automatically instantiate the RFR cursor when the criteria match

### A few notes on the design

The new CursorBasedCheckpointReader - Something that is really nice about this is that it is both low-code and Python stream agnostic. So long as a stream's implements the newly promoted Cursor concept, we can now avoid the checkpoint reader needing to scale w/ the complexities of state

I've tried to leverage the existing paginator logic as much as possible and avoid making the retriever and cursor have to re-implement behavior that already exists. However, it did require changes to allow reseting to a new value since there was no way to start pagination from anywhere but the start.

You'll also notice that the entire flow leverages the newly promoted `Cursor` and `StreamSlice` types that were recently promoted in https://github.com/airbytehq/airbyte/pull/38077 . This helps manage the complexity of the reader code when we can make clear assumptions on what is supplied to the source and passed back from the cursor.

## Review guide

1. `resumable_full_refresh_cursor.py`
2. `cursor_pagination_strategy.py`
3. `page_increment.py`
4. `offset_increment.py`
5. `default_paginator.py`
6. `simple_retriever.py`
7. `checkpoint_reader.py`
8. `model_to_component_transformer.py`

## User Impact

A few potential breaking changes:

- `PaginationStrategy.reset()` now can take in an optional parameter. This would affect custom pagination strategies, but a look through our repo only shows one example on `source-guardian-api` which is minimally used
- `Cursor.select_state()` has been added to the interface. This was originally unused by the PerPartitionCursor, but we will eventually need this for substream state. I wired most of it up, but custom components of Cursors/DatetimeBasedCursors need to implement this. It's unused by connectors right now
## Can this PR be safely reverted and rolled back?

- [ ] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/resumable_full_refresh_cursor.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        """
+        return True
+
+    def is_greater_than_or_equal(self, first: Record, second: Record) -> bool:
+        """
+        RFR record don't have ordering to be compared between one another.
+        """
+        return False
+
+    def select_state(self, stream_slice: Optional[StreamSlice] = None) -> Optional[StreamState]:
```
Comment (brianjlai) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/resumable_full_refresh_cursor.py:48]: This is a new interface method that we need in order to sanely read through state from the CheckpointReader. It isn't that useful for unnested streams, but this is critical for us to be able to sanely parse substream state as we iterate over parent record state in the CheckpointReader when we do the work

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/resumable_full_refresh_cursor.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        self._cursor = stream_slice.cursor_slice
+
+    def should_be_synced(self, record: Record) -> bool:
+        """
+        Unlike date-based cursors which filter out records outside slice boundaries, resumable full refresh records exist within pages
+        that don't have filterable bounds. We should always return them.
+        """
+        return True
+
+    def is_greater_than_or_equal(self, first: Record, second: Record) -> bool:
```
Comment (brianjlai) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/resumable_full_refresh_cursor.py:42]: This is the one field that really doesn't fit the existing cursor interface. We can default to false, and ultimately we don't even care about the record since we close the slice based on the page number.

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py [hunk 3]
```diff
@@ -71,6 +75,66 @@ def get_checkpoint(self) -> Optional[Mapping[str, Any]]:
...
+    def __init__(self, cursor: Cursor, stream_slices: Iterable[Optional[Mapping[str, Any]]], read_state_from_cursor: bool = False):
+        # The first attempt of an RFR stream has an empty {} incoming state, but should still make a first attempt to read records
+        # from the first page in next().
+        self._cursor = cursor
+        self._stream_slices = iter(stream_slices)
+        self._read_state_from_cursor = read_state_from_cursor
+        self._current_slice: Optional[StreamSlice] = None
+        self._finished_sync = False
+
+    def next(self) -> Optional[Mapping[str, Any]]:
```
Comment (brianjlai) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:95]: This method is a bit more complicated than it needs to be for streams because it can iterate in two dimensions: over incoming static set of slices and dynamically based on the current cursor's stream state.

Since this is only scoped to full refresh streams (not substreams  nor incremental), there should only be one static slice to loop over). But we support it regardless and positions us better for substream low-code RFR when we prioritize it.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_stream.py [hunk 4]
```diff
@@ -157,3 +159,8 @@ def state_checkpoint_interval(self) -> Optional[int]:
             important state is the one at the beginning of the slice
         """
         return None
+
+    def get_cursor(self) -> Optional[Cursor]:
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_stream.py:163]: more evidence the cursor belongs to the stream, not the retriever

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py [hunk 5]
```diff
@@ -668,6 +674,10 @@ def _merge_stream_slicers(self, model: DeclarativeStreamModel, config: Config) -
             )
         elif model.incremental_sync:
             return self._create_component_from_model(model=model.incremental_sync, config=config) if model.incremental_sync else None
+        elif hasattr(model.retriever, "paginator") and model.retriever.paginator and not stream_slicer:
+            # To incrementally deliver RFR for low-code we're first implementing this for streams that do not use
+            # nested state like substreams or those using list partition routers
+            return ResumableFullRefreshCursor(parameters={})
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py:680]: 👍 
Is there a follow up issue to support substreams?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py [hunk 6]
```diff
@@ -68,6 +71,7 @@ def __post_init__(self, parameters: Mapping[str, Any]) -> None:
         self._last_record: Optional[Record] = None
         self._parameters = parameters
         self._name = InterpolatedString(self._name, parameters=parameters) if isinstance(self._name, str) else self._name
+        self._synced_partitions: MutableMapping[Any, bool] = dict()
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py:74]: can you add a comment clarifying what the `True` means here? looking at a the code, it seems to mean that we started syncing a records from a partition.

It's also not obvious that this is only used for RFR.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py [hunk 7]
```diff
@@ -324,19 +346,42 @@ def read_records(
...
-            yield stream_data
+        if self.cursor and isinstance(self.cursor, ResumableFullRefreshCursor):
+            stream_state = self.state
 
-        if self.cursor:
-            self.cursor.close_slice(_slice, most_recent_record_from_slice)
+            # Before syncing the RFR stream, we check if the job's prior attempt was successful and don't need to fetch more records
+            # The platform deletes stream state for full refresh streams before starting a new job, so we don't need to worry about
+            # this value existing for the initial attempt
+            if stream_state.get(FULL_REFRESH_SYNC_COMPLETE_KEY):
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py:356]: thinking out loud: the fact that there are essentially two different code paths that are mutually exclusive for a given stream makes me think there's an opportunity to introduce a new abstraction where one implementation would be RFR and the other the standard `read_records`. 

I'd love to simplify this, but I'm not sure what the right abstraction is yet

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py (hunk 3 above)
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:95]: can you add a docstring describing the algorithm in English?

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py [hunk 8]
```diff
@@ -71,6 +75,66 @@ def get_checkpoint(self) -> Optional[Mapping[str, Any]]:
...
 
+class CursorBasedCheckpointReader(CheckpointReader):
+    """
+    CursorBasedCheckpointReader is used by streams that implement a Cursor in order to manage state. This allows the checkpoint
+    reader to delegate the complexity of fetching state to the cursor and focus on the iteration over a stream's partitions.
+    Right now only low-code connectors provide cursor implementations, but the logic is extensible to any stream that adheres
+    to the Cursor interface.
+    """
+
+    def __init__(self, cursor: Cursor, stream_slices: Iterable[Optional[Mapping[str, Any]]], read_state_from_cursor: bool = False):
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:86]: can you add a comment explaining when a user of this class how to configure `read_state_from_cursor`?

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py [hunk 9]
```diff
@@ -71,6 +75,66 @@ def get_checkpoint(self) -> Optional[Mapping[str, Any]]:
...
+        self._current_slice: Optional[StreamSlice] = None
+        self._finished_sync = False
+
+    def next(self) -> Optional[Mapping[str, Any]]:
+        try:
+            if self._current_slice is None:
+                self._current_slice = self._get_next_slice()
+                return self._current_slice
+            if self._read_state_from_cursor:
+                state_for_slice = self._cursor.select_state(self._current_slice.get("partition"))
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:101]: nit: I think it's preferable to call `self._current_slice.partition` instead of get as slowly gets us away from thinking of the slice as an arbitrary dict.

**Review by girarda — APPROVED** [review]
I'm a fan of this! This is a pretty big change but the abstractions make it manageable. ✅ 

I left a few comments but it's mostly just request for a few more comments. :shipit:

**Comment by brianjlai** [comment]
Ran regression tests successfully against 3 sources/connections:

- source-greenhouse - https://github.com/airbytehq/airbyte/actions/runs/9166779354/job/25202834408
- source-confluence (for offset increment) - https://github.com/airbytehq/airbyte/actions/runs/9155901010/job/25169175081
- source-freshdesk - https://github.com/airbytehq/airbyte/actions/runs/9167188398/job/25203934173

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py [hunk 10]
```diff
@@ -71,6 +75,66 @@ def get_checkpoint(self) -> Optional[Mapping[str, Any]]:
...
+                if state_for_slice == {"__ab_full_refresh_sync_complete": True}:
+                    self._current_slice = self._get_next_slice()
+                else:
+                    self._current_slice = StreamSlice(cursor_slice=state_for_slice or {}, partition=self._current_slice.partition)
+            else:
+                # Unlike RFR cursors that iterate dynamically based on how stream state is updated, most cursors operate on a
+                # fixed set of slices determined before reading records. They should just iterate to the next slice
+                self._current_slice = self._get_next_slice()
```
Comment (erohmensing) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:109]: Seeing a parallel between partition generators that generate partitions in advance and the fact that we were talking about creating partitions dynamically based on whether a next page cursor exists 🤔 just an observation

File: airbyte-cdk/python/unit_tests/sources/streams/test_streams_core.py [hunk 11]
```diff
@@ -215,8 +215,9 @@ def test_as_airbyte_stream_incremental(mocker):
         default_cursor_field=["test_cursor"],
         source_defined_cursor=True,
         source_defined_primary_key=[["primary_key"]],
+        is_resumable=True,
     )
-    assert exp == airbyte_stream
+    assert airbyte_stream == exp
```
Comment (erohmensing) [inline airbyte-cdk/python/unit_tests/sources/streams/test_streams_core.py:220]: ❤️

**Review by erohmensing — APPROVED** [review]

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py (hunk 5 above)
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py:680]: yep I have it filed here when the original spec was written: https://github.com/airbytehq/airbyte-internal-issues/issues/7528

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py (hunk 6 above)
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py:74]: good suggestion, will add!

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py (hunk 7 above)
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py:356]: Yeah I would agree with that. Its kind of our way of not needing to re-implement `read_records()` like we do for Python sources. It's not immediately clear at what level the abstraction needs to sit at yet. But I can't imagine pulling this out into its own abstraction later will be much of a lift.

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py (hunk 3 above)
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:95]: done

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py (hunk 9 above)
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:101]: actually i realize it should just be `self._current_slice` since `select_state()` takes in the slice itself, not the partition. But regardless this is fixed and we don't reference the arbitrary dict. Thanks!

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/strategies/offset_increment.py [hunk 12]
```diff
@@ -66,8 +66,13 @@ def next_page_token(self, response: requests.Response, last_page_size: int, last
...
 
-    def reset(self) -> None:
-        self._offset = 0
+    def reset(self, reset_value: Optional[Any] = None) -> None:
+        if reset_value is None:
+            self._offset = 0
+        elif not isinstance(reset_value, int):
+            raise ValueError(f"Reset value {reset_value} for OffsetIncrement pagination strategy was not an integer")
+        else:
+            self._offset = reset_value
```
[the remark this answers is not in the document]
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/strategies/offset_increment.py:75]: fixed!

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py [hunk 13]
```diff
@@ -301,6 +305,26 @@ def _read_pages(
...
+        stream_state: Mapping[str, Any],
+        stream_slice: StreamSlice,
+    ) -> Iterable[StreamData]:
+        response = self._fetch_next_page(stream_state, stream_slice)
+        yield from records_generator_fn(response)
+
+        if not response:
+            next_page_token: Mapping[str, Any] = {FULL_REFRESH_SYNC_COMPLETE_KEY: True}
+        else:
+            next_page_token = self._next_page_token(response) or {FULL_REFRESH_SYNC_COMPLETE_KEY: True}
```
[the remark this answers is not in the document]
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py:320]: yeah, i just ended up using similar logic that a non-existent response was the end of pagination which our existing pagination implementation does

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py [hunk 14]
```diff
@@ -404,6 +449,11 @@ def _parse_records(
     def must_deduplicate_query_params(self) -> bool:
         return True
 
+    @staticmethod
+    def _to_partition_key(to_serialize: Any) -> str:
+        # separators have changed in Python 3.4. To avoid being impacted by further change, we explicitly specify our own value
+        return json.dumps(to_serialize, indent=None, separators=(",", ":"), sort_keys=True)
+
```
[the remark this answers is not in the document]
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py:456]: no idea haha? 
I copied it from the per partition implementation here: https://github.com/airbytehq/airbyte/blob/8fdd9818ec16e653ba3dd2b167a74b7c07459861/airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py#L21-L24

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py [hunk 15]
```diff
@@ -324,19 +346,42 @@ def read_records(
...
+            # this value existing for the initial attempt
+            if stream_state.get(FULL_REFRESH_SYNC_COMPLETE_KEY):
+                return
+            cursor_value = stream_state.get("next_page_token")
+
+            # The first attempt to read a page for the current partition should have the paginator reset to the current
+            # cursor state which is initially set to the incoming state from the platform
+            partition_key = self._to_partition_key(_slice.partition)
+            if partition_key not in self._synced_partitions:
+                self._synced_partitions[partition_key] = True
```
[the remark this answers is not in the document]
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py:364]: Hmmm let me try to articulate it in an example, I think you have it mostly right. For our purposes yes lets just stick with one partition, although it is designed to work for parent stream records.

On the first attempt of the job, `_synced_partitions` is empty. So on the first iteration we pass the check since we know we haven't synced this partition yet. And we reset the value back to 0. It was already that, but we just reset it to itself. Then we continue paginating and we never reset because the partition is in the map. Now let's say we fail on page 10 and we've written state up to this point to page 9.

On the second attempt of the job `_synced_partitions` is empty so we reset this to the previous state of page 9. We then start paginating which has now been reset to page 9 until we complete.

On the next full refresh job, no state comes in so we paginate beginning at 0.

Does that help clarify things or did it just make it more confusing?

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py [hunk 16]
```diff
@@ -71,6 +75,66 @@ def get_checkpoint(self) -> Optional[Mapping[str, Any]]:
         return self._state
 
 
+class CursorBasedCheckpointReader(CheckpointReader):
+    """
+    CursorBasedCheckpointReader is used by streams that implement a Cursor in order to manage state. This allows the checkpoint
+    reader to delegate the complexity of fetching state to the cursor and focus on the iteration over a stream's partitions.
+    Right now only low-code connectors provide cursor implementations, but the logic is extensible to any stream that adheres
+    to the Cursor interface.
```
[the remark this answers is not in the document]
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:83]: I would say it is for the current type of Cursor we've promoted up. We're in a messy spot where we have a couple of different cursor interfaces though which is not great. But I will clarify this comment since you are right that it's confusing

File: airbyte-cdk/python/airbyte_cdk/sources/streams/core.py [hunk 17]
```diff
@@ -384,6 +385,14 @@ def get_updated_state(
...
 
+    def get_cursor(self) -> Optional[Cursor]:
+        """
+        A Cursor is an interface that a stream can implement to manage how its internal state is read and updated while
+        reading records. Historically, Python connectors had no concept of a cursor to manage state. Python streams need
+        need to define a cursor implementation and override this method to manage state through a Cursor.
+        """
+        return None
+
     def _get_checkpoint_reader(
```
[the remark this answers is not in the document]
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/core.py:396]: yeah i'm not sure. I did fiddle around with it a bit, but it's kinda hard to streamline it because the cursor reader needs the checkpoint reader. It is pretty hard to follow, but I think that's just a hazard of us having so many sync modes and flavors of CDK. :/ i might leave it as is for now since I don't have a good way to minimize the complexity and I tried to just keep it isolated with the `_get_checkpoint_reader()` function

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (erohmensing) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py:459]: Oh lol, no worries then! Same question for original author but 🤷🏻

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (erohmensing) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py]: Thank you!

File: airbyte-cdk/python/airbyte_cdk/sources/streams/core.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (erohmensing) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/core.py:396]: Fair enough, let's re-look at it when we've narrowed it down a bit 👍🏻

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (erohmensing) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py:367]: >  Then we continue paginating and we never reset because the partition is in the map.

I think this is what was confusing me, because `synced` to me indicates that we're _done_ syncing it, as in something that would be set at the end instead of the beginning. What about `started_partitions` since the act of resetting to the initial state is associated with starting to sync a partition? Or to Alex's point about it not being obvious, maybe `_rfr_initialized_partitions`?

> On the second attempt of the job _synced_partitions is empty so we reset this to the previous state of page 9. We then start paginating which has now been reset to page 9 until we complete.

This is definitely helpful on the logic side! Just the naming is caught for me now but I understand how its implemented.
